File: hook/collect-session-stats.py
#!/usr/bin/env python3
"""
Collect Claude Code session stats and append to ~/.claude/usage-log.jsonl.
Triggered by SessionEnd hook. Receives JSON payload via stdin.

Pricing table last updated: 2026-05-04
https://www.anthropic.com/pricing
"""

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

# Pricing per million tokens (USD) — update when Anthropic changes pricing
MODEL_PRICING = {
    "claude-opus-4-7": {
        "input": 15.00,
        "output": 75.00,
        "cache_write": 18.75,
        "cache_read": 1.50,
    },
    "claude-sonnet-4-6": {
        "input": 3.00,
        "output": 15.00,
        "cache_write": 3.75,
        "cache_read": 0.30,
    },
    "claude-haiku-4-5-20251001": {
        "input": 0.80,
        "output": 4.00,
        "cache_write": 1.00,
        "cache_read": 0.08,
    },
    "claude-haiku-4-5": {
        "input": 0.80,
        "output": 4.00,
        "cache_write": 1.00,
        "cache_read": 0.08,
    },
}

USAGE_LOG = Path.home() / ".claude" / "usage-log.jsonl"


def compute_cost(model: str, input_tokens: int, output_tokens: int,
                 cache_write_tokens: int, cache_read_tokens: int) -> tuple[float, bool]:
    """Returns (cost_usd, was_priced). was_priced=False means model not in table."""
    pricing = MODEL_PRICING.get(model)
    if not pricing:
        return 0.0, False
    cost = (
        input_tokens * pricing["input"] / 1_000_000
        + output_tokens * pricing["output"] / 1_000_000
        + cache_write_tokens * pricing["cache_write"] / 1_000_000
        + cache_read_tokens * pricing["cache_read"] / 1_000_000
    )
    return cost, True


def parse_transcript(transcript_path: str) -> dict:
    path = Path(transcript_path)
    if not path.exists():
        return {}

    tokens_by_model: dict[str, dict] = {}
    unpriced_models: set = set()
    timestamps: list[str] = []
    first_message = {}
    message_count = 0
    tool_call_count = 0

    with path.open() as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            try:
                entry = json.loads(line)
            except json.JSONDecodeError:
                continue

            ts = entry.get("timestamp")
            if ts:
                timestamps.append(ts)

            entry_type = entry.get("type")

            if entry_type == "user" and not first_message:
                first_message = entry

            if entry_type == "assistant":
                message_count += 1
                msg = entry.get("message", {})
                model = msg.get("model", "")

                # Skip synthetic/error messages (zero tokens, not real API calls)
                if model == "<synthetic>" or not model:
                    continue

                usage = msg.get("usage", {})
                input_tok = usage.get("input_tokens", 0)
                output_tok = usage.get("output_tokens", 0)
                cache_write = usage.get("cache_creation_input_tokens", 0)
                cache_read = usage.get("cache_read_input_tokens", 0)

                if model not in tokens_by_model:
                    tokens_by_model[model] = {
                        "input_tokens": 0,
                        "output_tokens": 0,
                        "cache_write_tokens": 0,
                        "cache_read_tokens": 0,
                    }

                tokens_by_model[model]["input_tokens"] += input_tok
                tokens_by_model[model]["output_tokens"] += output_tok
                tokens_by_model[model]["cache_write_tokens"] += cache_write
                tokens_by_model[model]["cache_read_tokens"] += cache_read

                # Count tool calls from content blocks
                for block in msg.get("content", []):
                    if isinstance(block, dict) and block.get("type") == "tool_use":
                        tool_call_count += 1

    # Compute cost per model
    total_cost = 0.0
    models_detail = {}
    for model, tok in tokens_by_model.items():
        cost, was_priced = compute_cost(
            model,
            tok["input_tokens"],
            tok["output_tokens"],
            tok["cache_write_tokens"],
            tok["cache_read_tokens"],
        )
        if not was_priced:
            unpriced_models.add(model)
        total_cost += cost
        models_detail[model] = {**tok, "cost_usd": round(cost, 6)}

    # Session duration
    duration_ms = None
    session_start = None
    session_end = None
    if len(timestamps) >= 2:
        try:
            t0 = datetime.fromisoformat(timestamps[0].replace("Z", "+00:00"))
            t1 = datetime.fromisoformat(timestamps[-1].replace("Z", "+00:00"))
            duration_ms = int((t1 - t0).total_seconds() * 1000)
            session_start = timestamps[0]
            session_end = timestamps[-1]
        except ValueError:
            pass
    elif timestamps:
        session_start = timestamps[0]
        session_end = timestamps[0]

    return {
        "session_start": session_start,
        "session_end": session_end,
        "duration_ms": duration_ms,
        "message_count": message_count,
        "tool_call_count": tool_call_count,
        "models": models_detail,
        "total_cost_usd": round(total_cost, 6),
        "unpriced_models": sorted(unpriced_models),
        "cwd": first_message.get("cwd"),
        "git_branch": first_message.get("gitBranch"),
        "version": first_message.get("version"),
    }


def main():
    try:
        payload = json.loads(sys.stdin.read())
    except (json.JSONDecodeError, OSError):
        sys.exit(0)

    session_id = payload.get("session_id", "")
    transcript_path = payload.get("transcript_path", "")

    if not transcript_path:
        sys.exit(0)

    stats = parse_transcript(transcript_path)
    if not stats.get("models"):
        sys.exit(0)

    record = {
        "session_id": session_id,
        "recorded_at": datetime.now(timezone.utc).isoformat(),
        **stats,
    }

    with USAGE_LOG.open("a") as f:
        f.write(json.dumps(record) + "\n")


if __name__ == "__main__":
    main()

File: src/collect_session_stats.rs
use std::collections::{HashMap, HashSet};
use std::fs::{File, OpenOptions};
use std::io::{self, BufRead, BufReader, Read, Write};
use std::path::PathBuf;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::Value;

// (input, output, cache_write, cache_read) per million tokens (USD)
// Last updated: 2026-06-13  https://www.anthropic.com/pricing
fn model_price(model: &str) -> Option<(f64, f64, f64, f64)> {
    match model {
        "claude-fable-5" => Some((10.00, 50.00, 12.50, 1.00)),
        "claude-opus-4-8" => Some((5.00, 25.00, 6.25, 0.50)),
        "claude-opus-4-7" => Some((15.00, 75.00, 18.75, 1.50)),
        "claude-sonnet-4-6" => Some((3.00, 15.00, 3.75, 0.30)),
        "claude-haiku-4-5-20251001" | "claude-haiku-4-5" => Some((0.80, 4.00, 1.00, 0.08)),
        _ if model.starts_with("claude-fable-5") => Some((10.00, 50.00, 12.50, 1.00)),
        _ if model.starts_with("claude-opus-4-8") => Some((5.00, 25.00, 6.25, 0.50)),
        _ if model.starts_with("claude-opus-4-7") => Some((15.00, 75.00, 18.75, 1.50)),
        _ if model.starts_with("claude-sonnet-4-6") => Some((3.00, 15.00, 3.75, 0.30)),
        _ if model.starts_with("claude-haiku-4-5") => Some((0.80, 4.00, 1.00, 0.08)),
        _ => None,
    }
}

#[derive(Deserialize)]
struct HookPayload {
    session_id: Option<String>,
    transcript_path: Option<String>,
}

#[derive(Deserialize)]
struct TranscriptEntry {
    #[serde(rename = "type")]
    entry_type: Option<String>,
    timestamp: Option<String>,
    message: Option<AssistantMessage>,
    cwd: Option<String>,
    #[serde(rename = "gitBranch")]
    git_branch: Option<String>,
    version: Option<String>,
}

#[derive(Deserialize)]
struct AssistantMessage {
    model: Option<String>,
    usage: Option<TokenUsage>,
    content: Option<Vec<Value>>,
}

#[derive(Deserialize, Default)]
struct TokenUsage {
    input_tokens: Option<u64>,
    output_tokens: Option<u64>,
    cache_creation_input_tokens: Option<u64>,
    cache_read_input_tokens: Option<u64>,
}

#[derive(Serialize, Default)]
struct ModelTokens {
    input_tokens: u64,
    output_tokens: u64,
    cache_write_tokens: u64,
    cache_read_tokens: u64,
    cost_usd: f64,
}

#[derive(Serialize)]
struct SessionRecord {
    session_id: String,
    recorded_at: String,
    session_start: Option<String>,
    session_end: Option<String>,
    duration_ms: Option<i64>,
    message_count: u64,
    tool_call_count: u64,
    models: HashMap<String, ModelTokens>,
    total_cost_usd: f64,
    unpriced_models: Vec<String>,
    cwd: Option<String>,
    git_branch: Option<String>,
    version: Option<String>,
}

fn round6(v: f64) -> f64 {
    (v * 1_000_000.0).round() / 1_000_000.0
}

fn parse_ts(s: &str) -> Option<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(&s.replace('Z', "+00:00"))
        .ok()
        .map(|dt| dt.with_timezone(&Utc))
}

fn main() {
    let mut input = String::new();
    if io::stdin().read_to_string(&mut input).is_err() {
        return;
    }

    let payload: HookPayload = match serde_json::from_str(&input) {
        Ok(p) => p,
        Err(_) => return,
    };

    let session_id = payload.session_id.unwrap_or_default();
    let transcript_path = match payload.transcript_path.as_deref() {
        Some(p) if !p.is_empty() => PathBuf::from(p),
        _ => return,
    };

    if !transcript_path.exists() {
        return;
    }

    let file = match File::open(&transcript_path) {
        Ok(f) => f,
        Err(_) => return,
    };

    let mut tokens_by_model: HashMap<String, ModelTokens> = HashMap::new();
    let mut unpriced: HashSet<String> = HashSet::new();
    let mut timestamps: Vec<String> = Vec::new();
    let mut first_user: Option<TranscriptEntry> = None;
    let mut message_count = 0u64;
    let mut tool_call_count = 0u64;

    for line in BufReader::new(file).lines() {
        let line = match line {
            Ok(l) => l,
            Err(_) => continue,
        };
        let line = line.trim().to_string();
        if line.is_empty() {
            continue;
        }
        let entry: TranscriptEntry = match serde_json::from_str(&line) {
            Ok(e) => e,
            Err(_) => continue,
        };

        if let Some(ts) = &entry.timestamp {
            timestamps.push(ts.clone());
        }

        match entry.entry_type.as_deref() {
            Some("user") => {
                if first_user.is_none() {
                    first_user = Some(entry);
                }
            }
            Some("assistant") => {
                message_count += 1;
                if let Some(msg) = &entry.message {
                    let model = msg.model.as_deref().unwrap_or("");
                    if model == "<synthetic>" || model.is_empty() {
                        continue;
                    }

                    let usage = msg.usage.as_ref();
                    let input = usage.and_then(|u| u.input_tokens).unwrap_or(0);
                    let output = usage.and_then(|u| u.output_tokens).unwrap_or(0);
                    let cache_write = usage.and_then(|u| u.cache_creation_input_tokens).unwrap_or(0);
                    let cache_read = usage.and_then(|u| u.cache_read_input_tokens).unwrap_or(0);

                    let tok = tokens_by_model.entry(model.to_string()).or_default();
                    tok.input_tokens += input;
                    tok.output_tokens += output;
                    tok.cache_write_tokens += cache_write;
                    tok.cache_read_tokens += cache_read;

                    if let Some(content) = &msg.content {
                        for block in content {
                            if block.get("type").and_then(Value::as_str) == Some("tool_use") {
                                tool_call_count += 1;
                            }
                        }
                    }
                }
            }
            _ => {}
        }
    }

    if tokens_by_model.is_empty() {
        return;
    }

    let mut total_cost = 0.0f64;
    for (model, tok) in &mut tokens_by_model {
        let cost = match model_price(model) {
            Some((ip, op, cwp, crp)) => {
                tok.input_tokens as f64 * ip / 1_000_000.0
                    + tok.output_tokens as f64 * op / 1_000_000.0
                    + tok.cache_write_tokens as f64 * cwp / 1_000_000.0
                    + tok.cache_read_tokens as f64 * crp / 1_000_000.0
            }
            None => {
                unpriced.insert(model.clone());
                0.0
            }
        };
        tok.cost_usd = round6(cost);
        total_cost += cost;
    }

    let (session_start, session_end, duration_ms) = if timestamps.len() >= 2 {
        let t0 = timestamps.first().unwrap();
        let t1 = timestamps.last().unwrap();
        let dur = match (parse_ts(t0), parse_ts(t1)) {
            (Some(dt0), Some(dt1)) => Some((dt1 - dt0).num_milliseconds()),
            _ => None,
        };
        (Some(t0.clone()), Some(t1.clone()), dur)
    } else {
        let ts = timestamps.into_iter().next();
        (ts.clone(), ts, None)
    };

    let (cwd, git_branch, version) = match first_user {
        Some(u) => (u.cwd, u.git_branch, u.version),
        None => (None, None, None),
    };

    let mut unpriced_models: Vec<String> = unpriced.into_iter().collect();
    unpriced_models.sort();

    let record = SessionRecord {
        session_id,
        recorded_at: Utc::now().to_rfc3339(),
        session_start,
        session_end,
        duration_ms,
        message_count,
        tool_call_count,
        models: tokens_by_model,
        total_cost_usd: round6(total_cost),
        unpriced_models,
        cwd,
        git_branch,
        version,
    };

    let log_path = dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join(".claude")
        .join("usage-log.jsonl");

    if let Ok(mut f) = OpenOptions::new().create(true).append(true).open(&log_path) {
        if let Ok(json) = serde_json::to_string(&record) {
            let _ = writeln!(f, "{}", json);
        }
    }
}
